docs/dev/use_pass_infra.rst

```diff
@@ -0,0 +1,117 @@
+passes as well), but also explains how we can build an optimization pipeline
+using ``Sequential`` associated with other types of passes.
+
+.. code:: python
+
+    # Create a simple Relay program.
+    shape = (1, 2, 3)
+    c_data = np.array(shape).astype("float32")
+    tp = relay.TensorType(shape, "float32")
+    c = relay.const(c_data)
+    x = relay.var("x", tp)
+    y = relay.add(c, c)
+    y = relay.multiply(y, relay.const(2, "float32"))
+    y = relay.add(x, y)
+    z = relay.add(y, c)
+    z1 = relay.add(y, c)
+    z2 = relay.add(z, z1)
+    func = relay.Function([x], z2)
+
+    # Customize the optimization pipeline.
+    seq = tvm.transform.Sequential([
+        relay.transform.InferType(),
+        relay.transform.FoldConstant(),
+        relay.transform.EliminateCommonSubexpr(),
+        relay.transform.AlterOpLayout()
+    ])
+
+    # Create a module to perform optimizations.
+    mod = relay.Module({"main": func})
+
+    # Users can disable any passes that they don't want to execute by providing
+    # a list, e.g. disabled_pass=["EliminateCommonSubexpr"].
+    with relay.build_config(opt_level=3):
+        with tvm.target.create("llvm"):
+            # Perform the optimizations.
+            mod = seq(mod)
+
+Debugging
+=========
+
+The pass infra provides a special pass (``PrintIR``) to dump the IR of the
+whole module after applying a certain pass. A slightly modified version of the
+sequential pass example could be like the following to enable IR dumping for
+``FoldConstant`` optimization.
+
+.. code:: python
+
+    seq = tvm.transform.Sequential([
+        relay.transform.InferType(),
+        relay.transform.FoldConstant(),
+        transform.PrintIR(),
+        relay.transform.EliminateCommonSubexpr(),
+        relay.transform.AlterOpLayout()
+    ])
+
+By inserting the ``PrintIR`` pass after ``FoldConstant``, the pass infra will
+dump out the module IR when ``FoldConstant`` is done. Users can plug in this
+pass after any pass they want to debug for viewing the optimization effect.
+
+There is a more flexible debugging mechanism also exposed by the build configuration
+object. One can pass a tracing function which can be used to execute arbitrary code
+before and/or after each pass. A tracing function will receive a ``IRModule``, ``PassInfo``,
+and a boolean indicating whether you are executing before, or after a pass.
+An example is below.
+
+.. code:: python
+
+    def print_ir(mod, info, is_before):
+        """Print the name of the pass, the IR, only before passes execute."""
+        if is_before:
+            print(f"Running pass: {}", info)
+            print(mod)
+
+    with relay.build_config(opt_level=3, trace=print_ir):
+            with tvm.target.create("llvm"):
+                # Perform the optimizations.
+                mod = seq(mod)
+
+
+For more pass infra related examples in Python and C++, please refer to
+`tests/python/relay/test_pass_manager.py`_ and
```

Comment:
Shall we put some examples in the tvm.ir.Sequential etc and refer to the API docs?

It seems to be not too useful to refer to the test code

Reply:
I realized that we have a tutorial already https://tvm.apache.org/docs/tutorials/dev/relay_pass_infra.html

We can probably just refactor this one and add it to how-to